### PR TITLE
support(transformer): add DigitalEdu tabular recipe

### PR DESCRIPTION
**Pull request by kujin66**
### Summary

Adds WinML 0.3.0 support for `SnowFlash383935/DigitalEduTransformers`, a remote-code binary tabular Transformer over seven numeric features. This Outcome-L2 change adds the shared `tabular-classification` task path, model/export wrapper, runtime pipeline, evaluator, tensor-comparison compatibility, and verified CPU fp32/fp16 recipes. Highest verified goal is L3 PASS: build, CPU perf, PyTorch parity, and a bounded two-sample functional Eval all pass on candidate `56f09d1d94290370387ce2cbdeeaa6a2e49524c5`.

### Model metadata

**What the model does:** Binary classification over seven numeric educational features. The checkpoint config declares `model_type="transformer"`, `architectures=["FleshkaTabularTransformer"]`, `input_dim=7`, and `pipeline_tag="tabular-classification"`.

**Primary user stories:** Convert and run the trained classifier through WinML/ONNX with a stable tensor contract, and evaluate it with the same zero-threshold boolean semantics used by the checkpoint's remote code.

**Supported tasks:** `tabular-classification`, with tensor input `features` and output `logits`.

**Model architecture:**

```text
FleshkaTabularTransformer
├── feature normalization (mean/std, 7 features)
├── input projection (7 -> 512)
├── Transformer encoder layer x 32
│   ├── multi-head self-attention (16 heads)
│   ├── feed-forward block
│   └── residual + LayerNorm
└── binary classification head (512 -> 1 logit)
```

- Source/confidence: checkpoint config, model card, and remote `model.py` for `SnowFlash383935/DigitalEduTransformers`; verified against the loaded `TabularTransformerWrapper` (100,881,921 parameters).

#### Baseline

The rerun started from WinML 0.3.0 at `origin/main@48bfd0f9`. Baseline config with `--task tabular-classification --trust-remote-code` could not resolve a supported task/export boundary, normal pipeline construction had no Transformers task for tabular classification, and `winml eval` had no task evaluator. The historical PR head was 96 commits behind and conflicted, so this was a FULL-RERUN rather than a mechanical rebase. Before shipment, `origin/main` moved to `2b9ec0e9`; its complete delta only changed `docs/commands/sys.md`, `src/winml/modelkit/commands/sys.py`, and corresponding sys tests. The candidate was rebased without conflict, the patch remained equivalent on owned paths, and final-head focused tests were rerun.

Starting auto-config behavior: no usable recipe could be emitted because the task and model wrapper were unregistered. Optimum exposes a text-classification exporter boundary, but not this remote tabular model/task contract; WinML now maps the new task to that boundary only inside the per-architecture wrapper.

#### Goal

- **Effort:** L2, because support requires a model wrapper plus a new shared task family.
- **Goal ceiling:** L3, defined as L0 build/structure, L1 CPU perf, L2 PyTorch-vs-ONNX numeric comparison, and one semantically valid FP32 CPU functional smoke Eval.
- **Outcome:** L2 task-family support with CPU fp32/fp16 recipes and regression coverage.

#### Outcome

L0-L3 all PASS on final candidate `56f09d1d94290370387ce2cbdeeaa6a2e49524c5`. Coverage is full for the declared CPU target: both required precision tuples are freshly built and measured. No accelerator runtime support is claimed; `--ep all` below is static rule analysis only.

Shipped recipes:

- `examples/recipes/SnowFlash383935_DigitalEduTransformers/cpu/cpu/tabular-classification_fp32_config.json`
- `examples/recipes/SnowFlash383935_DigitalEduTransformers/cpu/cpu/tabular-classification_fp16_config.json`

Shipped capability includes the tabular task registry, remote-code-aware model loading, tensor-native serving pipeline, binary evaluator, raw-tensor comparison mapping, and localized Transformer export compatibility. No new methodology friction was observed beyond cases already covered by the current skill.

#### Per-EP/device/precision results and Functional smoke Eval

| Tier | EP / Device | Precision | Verdict | Mean | p50 | Throughput | RAM delta |
|---|---|---|---|---:|---:|---:|---:|
| L0 | CPUExecutionProvider / cpu | fp32 | PASS; build 93.7 s; 385.2 MB weights | - | - | - | - |
| L0 | CPUExecutionProvider / cpu | fp16 | PASS; build 102.7 s; 192.8 MB weights | - | - | - | - |
| L1 | CPUExecutionProvider / cpu | fp32 | PASS | 17.74 ms | 17.31 ms | 56.36 samples/s | +2.5 MB total |
| L1 | CPUExecutionProvider / cpu | fp16 | PASS | 18.86 ms | 18.88 ms | 53.03 samples/s | +1.3 MB total |
| L2 | CPUExecutionProvider / cpu | fp32 | PASS | - | - | - | cosine 1.0000; max-abs 0.0000 |
| L2 | CPUExecutionProvider / cpu | fp16 | PASS | - | - | - | cosine 1.0000; max-abs mean 0.0004, max 0.0008 |

Both artifacts load as ONNX IR 8/opset 17 with `features [1,7] float32 -> logits [1,1] float32`. FP32 has 393 FLOAT initializers and a 384.83 MB sidecar; FP16 has 393 FLOAT16 initializers and a 192.42 MB sidecar.

**Functional smoke Eval:** L3 PASS on final candidate SHA, FP32 CPU, using the two real examples published in the model card: `[1,1,1,150,2,0,0] -> False` and `[0,1,1,294,0,0,0] -> True`. The local Dataset contained exactly two selected and two processed rows, `--samples 2`, `--no-shuffle`, no label/prompt/beam/frame/crop/generation expansion, labels encoded as `[0,1]`, and predictions interpreted with the remote model's `logit > 0` binary rule. Raw metrics: accuracy `1.0`, F1 `1.0`, `num_samples=2`. This proves end-to-end evaluator operability only; it is not representative accuracy or a benchmark-quality claim. The former blocker was the missing tabular evaluator/pipeline and trust propagation, which this PR implements and regression-tests.

Quality gates on the candidate:

- `uv run ruff check src/ tests/`: PASS.
- `uv run mypy -p winml.modelkit`: PASS, 439 source files.
- `uv run pytest tests/unit/eval --tb=short --no-cov -m "not e2e and not npu and not gpu"`: 630 passed.
- Final-head focused Eval/pipeline regression set: 127 passed.
- Final-head models CI partition: 1537 passed, 6 skipped, 2 xfailed.
- `uv run pre-commit run insert-license --all-files`: PASS.

#### Delta

The baseline had no generated tabular recipe to diff. The shipped recipes therefore make the previously unresolved contract explicit: `loader.model_type=transformer`, `loader.model_class=TabularTransformerWrapper`, `loader.task=tabular-classification`, `loader.trust_remote_code=true`, opset 17, `features` float32 shape `[1,7]`, and output `logits`; the fp16 recipe additionally sets `quant.mode=fp16`. The production `examples/recipes/README.md` is unchanged.

**Bug fix explanation:** (a) Minimal triggers were config/build on the remote checkpoint, normal `create_pipeline("tabular-classification", ...)`, and `winml eval`; these respectively lacked a model/task boundary, a native Transformers pipeline, and an evaluator. (b) The root causes were missing shared task registration, missing trust propagation, an old remote class incompatible with Transformers 5 tied-weight metadata, PyTorch's fused MHA export path at opset 17, tensor comparison assuming a `ModelOutput` mapping, and a task wrapper override that added a required positional parameter relative to the generic base signature. (c) The fix adds `TabularTransformerWrapper`, `WinMLModelForTabularClassification`, `WinMLTabularClassificationEvaluator`, `_TabularClassificationPipeline`, task/schema registrations, trust forwarding, raw-tensor output naming, compatible config loading, and a generic `forward(**kwargs)` runtime signature; export temporarily disables the MHA fastpath only while ONNX export is active and restores it in `finally`. The runtime output uses an `Any` cast boundary because Transformers' legacy `FloatTensor` annotation is narrower than the real tensor returned by the ONNX session, avoiding a type-only runtime import. (d) Dispatch is based on registered task/model class and output shape, never the checkpoint ID; one-logit and two-logit binary outputs are both supported. (e) Existing tasks retain their mappings and defaults, remote code remains opt-in, the base runtime call contract is preserved, and the fastpath global is restored after successful or failed export. (f) Regression evidence includes the 630-test Eval partition, 1537-test models partition, 69-test runtime-focused set, full Ruff/mypy, both exact recipe builds, both CPU perf runs, both parity runs, and the two-row functional Eval above.

#### Analyze summary — component level and op level

Static rule analysis completed for both artifacts with `WINMLCLI_RULES_DIR` and `--ep all`; this is compatibility classification, not runtime execution.

**Component-level summary**

| Artifact | Architecture coverage | Mapping | Actionable EP findings |
|---|---|---|---|
| fp32 | normalization; projection; 32x attention/FFN; binary head | export tagged 2668/2668 nodes, all model-root fallback; final optimized graph 1287 nodes | TensorRT RTX, QNN GPU, OpenVINO GPU rules classify most types supported; conditional `Reshape` and `Softmax` remain unknown |
| fp16 | same regions plus FP16 boundary casts | export tagged 2668/2668 nodes, all model-root fallback; final graph 1289 nodes | same unknowns plus conditional `Cast`; no partial or unsupported type reported |

Mapping confidence is fallback-only because remote-module tracing captured 1/327 modules; node scope names still expose `input_proj`, `transformer/layers.0..31`, normalizations, and the head. This limitation is retained rather than presented as high-confidence per-module mapping.

**Op-level summary**

| Artifact | Graph | Dominant ops | EP roll-up |
|---|---|---|---|
| fp32 | 1287 ops / 14 types | Reshape 448; Transpose 256; Gemm 130; Gather 96; LayerNormalization 65 | TensorRT RTX/QNN GPU/OpenVINO GPU: supported types with conditional Reshape/Softmax unknown; CUDA/MIGraphX/TensorRT/DML have no rule classifications |
| fp16 | 1289 ops / 15 types | Reshape 448; Transpose 256; Gemm 130; Gather 96; LayerNormalization 65; Cast 2 | same, with conditional Cast also unknown |

No Analyze row reported an unsupported operator type. Rule-less rows are not runtime claims.

#### Reproduce commands

```powershell
$OUT = 'temp/digitaledu-repro'
uv run winml build -c examples\recipes\SnowFlash383935_DigitalEduTransformers\cpu\cpu\tabular-classification_fp32_config.json -m SnowFlash383935/DigitalEduTransformers -o $OUT\fp32 --trust-remote-code --rebuild
uv run winml build -c examples\recipes\SnowFlash383935_DigitalEduTransformers\cpu\cpu\tabular-classification_fp16_config.json -m SnowFlash383935/DigitalEduTransformers -o $OUT\fp16 --trust-remote-code --precision fp16 --rebuild
uv run winml perf -m $OUT\fp32\model.onnx --ep cpu --device cpu
uv run winml perf -m $OUT\fp16\model.onnx --ep cpu --device cpu
uv run python -c "from datasets import Dataset; Dataset.from_dict({'features':[[1,1,1,150,2,0,0],[0,1,1,294,0,0,0]],'label':[0,1]}).save_to_disk(r'$OUT\model-card-dataset')"
uv run winml eval -m $OUT\fp32\model.onnx --model-id SnowFlash383935/DigitalEduTransformers --task tabular-classification --dataset $OUT\model-card-dataset --samples 2 --no-shuffle --ep cpu --device cpu --trust-remote-code --output $OUT\fp32\eval_model_card.json
$env:WINMLCLI_RULES_DIR = '<populated-rules-directory>'
uv run winml analyze --model $OUT\fp32\model.onnx --ep all --output $OUT\fp32\analyze_all.json
uv run winml analyze --model $OUT\fp16\model.onnx --ep all --output $OUT\fp16\analyze_all.json
Remove-Item Env:WINMLCLI_RULES_DIR
```

`--trust-remote-code` is required because this checkpoint loads custom `auto_map` Python from its model repository.

### CODE REVIEW

**Comment by kujin66** [comment]
APPROVE (skill reviewer verdict)

Reviewed PR #1198 against the `adding-model-support` reviewer checklist.

- Shipment state: draft PR, `model-scale-by-skill` label present, head `ce190c164ea9b3112b028e354987a183a21d756d`.
- Diff scope: L2 Lane B task/architecture support. The diff contains the new tabular Transformer wrapper/export registration, task registry/inference mapping, CPU fp32/fp16 recipes, and targeted tests; `examples/recipes/README.md` is untouched.
- Required PR hierarchy: Summary, Model metadata, and Validation/support evidence are present with Baseline, Goal, Outcome, per-EP/device/precision results, Delta, Analyze summary, and Reproduce commands.
- Verification evidence: `--trust-remote-code` config/build path is documented; CPU fp32/fp16 L0 build, L1 perf, L2 PyTorch-vs-ONNX comparison, fp32/fp16 analyze, and L3 CLI blocker are recorded.
- Engineering review: wrapper exposes the trained remote model through tensor `features -> logits` without hardcoding a model id in shared code; OnnxConfig dynamic axes match the recipe; task boundary maps `tabular-classification` to Optimum `text-classification` only at the exporter edge.
- GitHub checks on reviewed SHA: `Analyze (Python)`, `lint`, `test (analyze)`, `test (commands)`, `test (models)`, `test (optim)`, `test (remaining)`, `CodeQL`, and `license/cla` all completed successfully.
- Review threads/comments: REST line comments empty; issue comments empty before this verdict; GraphQL reviewThreads empty.

No requested changes. Leave the PR in draft state per skill workflow.

File: src/winml/modelkit/models/winml/tabular_classification.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    import torch
+
+
+class WinMLModelForTabularClassification(WinMLPreTrainedModel):
+    """Run a tabular classifier whose ONNX input is named ``features``."""
+
+    def forward(  # type: ignore[override]
+        self,
+        features: torch.Tensor | np.ndarray,
+    ) -> SequenceClassifierOutput:
```
[the remark this answers is not in the document]
Comment (ssss141414) [reply src/winml/modelkit/models/winml/tabular_classification.py]: Fixed in 65934caa6af10b0f830f1a1155f2cf7deb23cdea. WinMLModelForTabularClassification.forward now preserves the generic base contract with forward(**kwargs), while _format_inputs still validates and formats the features input. Added a signature regression test; focused runtime tests (69), models partition (1537), Eval partition (630), Ruff, and mypy all pass.

File: src/winml/modelkit/models/winml/tabular_classification.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+from typing import TYPE_CHECKING, Any, cast
+
+from transformers.modeling_outputs import SequenceClassifierOutput
+
+from .base import WinMLPreTrainedModel
+
+
+if TYPE_CHECKING:
+    import torch
```
[the remark this answers is not in the document]
Comment (ssss141414) [reply src/winml/modelkit/models/winml/tabular_classification.py]: Fixed in 56f09d1d94290370387ce2cbdeeaa6a2e49524c5. Removed the type-only torch import and kept the real ONNX tensor behind a quoted Any cast at the Transformers legacy output annotation boundary. Ruff passes, mypy reports 0 issues in 439 files, focused runtime tests pass 69/69, models partition passes 1537 tests, and Eval passes 630 tests.

**Comment by ssss141414** [comment]
APPROVE (skill reviewer verdict)

Reviewed PR #1198 against the latest `adding-model-support` reviewer checklist at exact head `56f09d1d94290370387ce2cbdeeaa6a2e49524c5`.

- Shipment state: Draft PR with the `model-scale-by-skill` label; local and remote heads match.
- Scope: adds `SnowFlash383935/DigitalEduTransformers` support for `tabular-classification`, including current-loader compatibility, ONNX export/runtime registration, FP32/FP16 CPU recipes, bounded functional Eval, tensor comparison, and targeted tests.
- Functional smoke: two real model-card samples on FP32 CPU; accuracy `1.0`, F1 `1.0`.
- Parity: FP32 cosine `1.0`; FP16 cosine `1.0` with maximum absolute error `0.0007920`.
- Performance: FP32 mean/p50 `17.74/17.31 ms`, `56.36 samples/s`, RAM delta `+2.5 MB`; FP16 mean/p50 `18.86/18.88 ms`, `53.03 samples/s`, RAM delta `+1.3 MB`.
- Static analysis: FP32 `1287` operators / `14` types; FP16 `1289` operators / `15` types. Both artifacts preserve `features [1,7] -> logits [1,1]`.
- Local gates: Ruff passed; mypy passed for 439 files; models partition `1537 passed`; Eval partition `630 passed`; focused runtime tests passed.
- GitHub gates: all 9 checks succeeded on the reviewed SHA; no pending, failed, or cancelled checks.
- Feedback gate: both CodeQL review threads are resolved; unresolved review thread count is zero.
- Coverage accounting: declared CPU FP32 and FP16 targets are covered. No accelerator runtime claim is made.

No blocking findings. Leave the PR in Draft state per the skill workflow.
